Mathlib/Data/List/Infix.lean

```diff
@@ -70,6 +70,19 @@ theorem infix_rfl : l <:+: l :=
 theorem prefix_concat (a : α) (l) : l <+: concat l a := by simp
 #align list.prefix_concat List.prefix_concat
 
+theorem prefix_concat_iff {x y : List α} {a : α} : x <+: y ++ [a] ↔ x <+: y ∨ x = y ++ [a] := by
+  constructor
+  · intro ⟨z, h⟩; induction' z using List.reverseRecOn with z
+    · right
```

Comment:
You can simplify this slightly, using the `Or.inl` and `Or.inr` lemmas:
```lean
rw [List.append_nil] at h
exact Or.inr h
```
for the first one, and using `´Or.inl (full previous term)` for the second one.